smart_open/tests/test_smart_open.py

```diff
@@ -200,92 +201,173 @@ def test_http_pass(self):
+            return self.fobj.write(b)
 
-        if os.path.isfile(test_file):
-            os.unlink(test_file)
+        def close(self):
+            self.fobj.close()
 
-        responses.add(responses.GET, "http://127.0.0.1/data.xz", body=compressed_data, stream=True)
-        smart_open_object = smart_open.smart_open("http://127.0.0.1/data.xz")
+    def test_read(self):
+        file = self.IOReadWrapper()
+        text = 'Hello, world!'
+        file.fobj.write(text.encode())
+        file.fobj.seek(0)
+        with smart_open.smart_open(file, 'r') as sf:
+            data = sf.read()
+        self.assertEqual(data, text)
+        file = self.IOWriteWrapper()  # no `read()`
+        file.fobj.write(text.encode())
+        file.fobj.seek(0)
+        self.assertRaises(TypeError, lambda: smart_open.smart_open(file, 'r'))
 
-        # decompress the xz and get the same md5 hash
-        self.assertEqual(smart_open_object.read(), test_string)
+    def test_write(self):
+        file = self.IOWriteWrapper()
+        text = 'Hello, world!'
+        with smart_open.smart_open(file, 'w') as sf:
+            sf.write(text)
+            self.assertEqual(file.fobj.getvalue(), text.encode())
+        file = self.IOReadWrapper()  # no `write()`
+        self.assertRaises(TypeError, lambda: smart_open.smart_open(file, 'w'))
+
+    def test_read_write(self):
+        file = six.BytesIO()
+        text = 'Hello, world!'
+        file.write(text.encode())
+        file.seek(0)
+        with smart_open.smart_open(file, 'r+') as sf:
+            data = sf.read()
+            sf.write("hello")
+        self.assertEqual(data, text)
+
+        file = six.BytesIO()
+        file.write(text.encode())
+        file.seek(0)
+        with smart_open.smart_open(file, 'w+') as sf:
+            data = sf.read()
+            sf.write("hello")
+        self.assertEqual(data, text)
+
+        file = six.BytesIO()
+        file.write(text.encode())
+        file.seek(0)
+        with smart_open.smart_open(file, 'a+') as sf:
+            data = sf.read()
+            sf.write("hello")
+        self.assertEqual(data, text)
+
+        file = self.IOReadWrapper()  # no `write()`
+        self.assertRaises(TypeError, lambda: smart_open.smart_open(file, 'r+'))
+        self.assertRaises(TypeError, lambda: smart_open.smart_open(file, 'w+'))
+        self.assertRaises(TypeError, lambda: smart_open.smart_open(file, 'a+'))
+
+        file = self.IOWriteWrapper()  # no `read()`
+        self.assertRaises(TypeError, lambda: smart_open.smart_open(file, 'r+'))
+        self.assertRaises(TypeError, lambda: smart_open.smart_open(file, 'w+'))
+        self.assertRaises(TypeError, lambda: smart_open.smart_open(file, 'a+'))
+
+    def test_append(self):
+        file = self.IOWriteWrapper()
+        text = 'Hello, world!'
+        with smart_open.smart_open(file, 'a') as sf:
+            sf.write(text)
+            self.assertEqual(file.fobj.getvalue(), text.encode())
```

Comment:
Please move the assertion to outside of the context manager (here and elsewhere).

Reply:
That's intentional: the file object is closed in the context manager and `getvalue` raises an exception